src/Builder/DefaultDriverFactory.php

```diff
@@ -54,13 +55,21 @@ public function enableEnumSupport(bool $enableEnumSupport = true): void
         $this->enableEnumSupport = $enableEnumSupport;
     }
 
-    public function createDriver(array $metadataDirs, Reader $annotationReader): DriverInterface
+    public function createDriver(array $metadataDirs, ?Reader $annotationReader = null): DriverInterface
     {
+        if (empty($metadataDirs) && !interface_exists(Reader::class) && PHP_VERSION_ID < 80000) {
+            throw new RuntimeException(sprintf('To use "%s", either a list of metadata directories must be provided, the "doctrine/annotations" package installed, or use PHP 8.0 or later.', self::class));
+        }
+
+        $driver = new DriverChain();
+
         if (PHP_VERSION_ID >= 80000) {
-            $annotationReader = new AttributeDriver\AttributeReader($annotationReader);
+            $driver->addDriver(new AttributeDriver($this->propertyNamingStrategy, $this->typeParser, $this->expressionEvaluator));
         }
 
-        $driver = new AnnotationDriver($annotationReader, $this->propertyNamingStrategy, $this->typeParser);
+        if ($annotationReader) {
```

Comment:
I wonder is we should modify `ChainDriver` too?

Reply:
The instance at line 76 just outside this diff or something about the class itself?

There isn't a prepend method on the chain, so the implementation I have now ended up being the simplest one considering that when metadata paths are given, the factory pushes the YAML and XML drivers up in priority over the annotations driver (and the same goes for the default ordering in the bundle service definitions).